shared/engine/session.js

```diff
@@ -0,0 +1,125 @@
+  }
+
+  set id (sessionID: SessionID) { throw new Error("Can't set sessionID") }
+  get id (): SessionID { return this._id }
+
+  // Make a waiting handler for the request. We add additional data before calling the parent waitingHandler
+  // and do internal bookkeeping if the request is done
+  _makeWaitingHandler (isOutgoing: boolean, method: MethodKey, seqid: ?number) {
+    return (waiting: boolean) => {
+      rpcLog('engineInternal', 'waiting state change', this.id, waiting, method, this, seqid)
+      // Call the outer handler with all the params it needs
+      this._waitingHandler && this._waitingHandler(waiting, method, this._id)
+
+      // Request is finished, do cleanup
+      if (!waiting) {
+        const requests = isOutgoing ? this._outgoingRequests : this._incomingRequests
+        const idx = requests.findIndex(r => r.method === method)
+        if (idx !== -1) {
+          // Mark us as responded
+          if (seqid) {
+            this._seqIDResponded[String(seqid)] = true
+          }
+          // Remove from our list
+          requests.splice(idx, 1)
+        }
+      }
+    }
+  }
+
+  end (): void {
+    this._endHandler(this)
+  }
+
+  // Start the session normally. Tells engine we're done at the end
+  start (method: MethodKey, param: ?Object, callback: ?() => void) {
+    // When this request is done the session is done
+    const wrappedCallback = (...args) => {
+      callback && callback(...args)
+      this.end()
+    }
+
+    // Add the sessionID
+    const wrappedParam = {
+      ...param,
+      sessionID: this.id,
+    }
+
+    rpcLog('engineInternal', 'session start call', this.id, method, this)
+    const outgoingRequest = new OutgoingRequest(method, wrappedParam, wrappedCallback, this._makeWaitingHandler(true, method), this._invoke)
+    this._outgoingRequests.push(outgoingRequest)
+    outgoingRequest.send()
+  }
+
+  // We have an incoming call tied to a sessionID, called only by engine
+  incomingCall (method: MethodKey, param: Object, response: ?Object): boolean {
+    rpcLog('engineInternal', 'session incoming call', this.id, method, this, response)
+    const handler = this._incomingCallMap[method]
+
+    if (!handler) {
+      return false
+    }
+
+    if (response && response.seqid) {
+      this._seqIDResponded[String(response.seqid)] = false
+    }
+
+    const waitingHandler = this._makeWaitingHandler(false, method, response && response.seqid)
+    const incomingRequest = new IncomingRequest(method, param, response, waitingHandler, handler)
+    this._incomingRequests.push(incomingRequest)
+    incomingRequest.handle()
+
+    return true
+  }
+
+  // Tell engine if we can handle the cancelled call
+  hasSeqID (seqID: number) {
+    if (__DEV__) {
+      if (this._seqIDResponded.hasOwnProperty(String(seqID))) {
+        console.log('Cancelling seqid found, current session state', this)
+      }
+    }
+    return this._seqIDResponded.hasOwnProperty(String(seqID))
```

Comment:
what if seqIDResponded is `true`? that means we've already handled the call and the service wants us to cancel it?